File: plugins/FC_XenForo2/upload/src/addons/ForumCopilot/Result/FCPostResult.php
<?php

namespace ForumCopilot\Result;

/**
 * ForumCopilot Reply Post Result
 * Maps from ReplyPostData_Output
 */
class FCReplyPostResult extends FCResult
{
    public $postId;
    public $state;
    public $postContent;
    public $canEdit;
    public $canDelete;
    public $canReport;

    public function __construct($result = true, $resultText = 'success', $postId = null, $state = 0, $postContent = null, $canEdit = false, $canDelete = false, $canReport = false)
    {
        parent::__construct($result, $resultText);
        $this->postId = $postId;
        $this->state = $state;
        $this->postContent = $postContent;
        $this->canEdit = $canEdit;
        $this->canDelete = $canDelete;
        $this->canReport = $canReport;
    }

    public function toArray()
    {
        $data = parent::toArray();
        if ($this->postId !== null) $data['postId'] = $this->postId;
        $data['state'] = $this->state;
        if ($this->postContent !== null) $data['postContent'] = $this->postContent;
        $data['canEdit'] = $this->canEdit;
        $data['canDelete'] = $this->canDelete;
        $data['canReport'] = $this->canReport;
        return $data;
    }
}

/**
 * ForumCopilot Report Post Result
 * Maps from ReportPostData_Output
 */
class FCReportPostResult extends FCResult
{
    public function __construct($result = true, $resultText = 'success')
    {
        parent::__construct($result, $resultText);
    }

    public function toArray()
    {
        return parent::toArray();
    }
}

/**
 * ForumCopilot Quote Post Result
 * Maps from QuotePostData_Output
 */
class FCQuotePostResult extends FCResult
{
    public $quoteContent;

    public function __construct($result = true, $resultText = 'success', $quoteContent = null)
    {
        parent::__construct($result, $resultText);
        $this->quoteContent = $quoteContent;
    }

    public function toArray()
    {
        $data = parent::toArray();
        if ($this->quoteContent !== null) $data['quoteContent'] = $this->quoteContent;
        return $data;
    }
}

/**
 * ForumCopilot Raw Post Result
 * Maps from RawPostData_Output
 */
class FCRawPostResult extends FCResult
{
    public $postContent;
    public $postTitle;
    public $canEditTitle;
    public $prefixId;
    public $requirePrefix;
    public $prefixes;
    public $attachments;

    public function __construct($result = true, $resultText = 'success', $postContent = null, $postTitle = null, $canEditTitle = false, $prefixId = null, $requirePrefix = false, $prefixes = [], $attachments = [])
    {
        parent::__construct($result, $resultText);
        $this->postContent = $postContent;
        $this->postTitle = $postTitle;
        $this->canEditTitle = $canEditTitle;
        $this->prefixId = $prefixId;
        $this->requirePrefix = $requirePrefix;
        $this->prefixes = $prefixes;
        $this->attachments = $attachments;
    }

    public function toArray()
    {
        $data = parent::toArray();
        if ($this->postContent !== null) $data['postContent'] = $this->postContent;
        if ($this->postTitle !== null) $data['postTitle'] = $this->postTitle;
        $data['canEditTitle'] = $this->canEditTitle;
        // Always include prefixId, even if null
        $data['prefixId'] = $this->prefixId;
        $data['requirePrefix'] = $this->requirePrefix;
        if (!empty($this->prefixes)) {
            $data['prefixes'] = array_map(function($prefix) {
                return is_object($prefix) ? $prefix->toArray() : $prefix;
            }, $this->prefixes);
        } else {
            $data['prefixes'] = [];
        }
        // Convert attachments to arrays if they are FCAttachment objects
        $attachmentsArray = [];
        if (is_array($this->attachments)) {
            foreach ($this->attachments as $attachment) {
                if (is_object($attachment) && method_exists($attachment, 'toArray')) {
                    $attachmentsArray[] = $attachment->toArray();
                } else {
                    $attachmentsArray[] = $attachment;
                }
            }
        }
        $data['attachments'] = $attachmentsArray;
        return $data;
    }
}

/**
 * ForumCopilot Save Raw Post Result
 * Maps from SaveRawPostData_Output
 */
class FCSaveRawPostResult extends FCResult
{
    public $postContent;

    public function __construct($result = true, $resultText = 'success', $postContent = null)
    {
        parent::__construct($result, $resultText);
        $this->postContent = $postContent;
    }

    public function toArray()
    {
        $data = parent::toArray();
        if ($this->postContent !== null) $data['postContent'] = $this->postContent;
        return $data;
    }
}

/**
 * ForumCopilot Thread Result
 * Maps from ThreadData_Output
 * Contains all FCTopic properties plus posts array
 */
class FCThreadResult extends FCResult
{
    // FCTopic properties
    public $id;
    public $title;
    public $forumId;
    public $forumName;
    public $authorId;
    public $authorName;
    public $authorUserType;
    public $timestamp;
    public $prefix;
    public $authorIconUrl;
    public $replyCount;
    public $viewCount;
    public $hasNewPosts;
    public $isClosed;
    public $isSubscribed;
    public $canSubscribe;
    public $url;
    public $shortContent;
    public $participatedUserIds;
    public $isPinned;
    public $isAnnouncement;
    public $isStickySource;
    public $canRename;
    public $canDelete;
    public $canClose;
    public $canApprove;
    public $canStick;
    public $canMove;
    public $canMerge;
    public $canBan;
    public $canReply;
    public $canReport;
    public $canUpload;
    public $isBanned;
    public $isApproved;
    public $isDeleted;
    public $isMoved;
    public $isMerged;
    public $realTopicId;
    public $canLike;
    public $isLiked;
    public $likeCount;
    public $hasPoll;
    public $poll;

    // Thread-specific properties
    public $totalPostNum;
    public $posts;

    public function __construct($result = true, $resultText = 'success', $totalPostNum = 0, $posts = [], $topicData = [])
    {
        parent::__construct($result, $resultText);
        $this->totalPostNum = $totalPostNum;
        $this->posts = $posts;
        
        // Set all topic properties from $topicData
        if (is_array($topicData)) {
            foreach ($topicData as $key => $value) {
                if (property_exists($this, $key)) {
                    $this->$key = $value;
                }
            }
        }
    }

    public function toArray()
    {
        $data = parent::toArray();
        $data['totalPostNum'] = $this->totalPostNum;
        $data['posts'] = array_map(function($post) {
            return is_object($post) ? $post->toArray() : $post;
        }, $this->posts);
        
        // Add all topic properties (include hasPoll and poll for thread responses)
        $topicProps = ['id', 'title', 'forumId', 'forumName', 'authorId', 'authorName', 'authorUserType',
                      'timestamp', 'prefix', 'authorIconUrl', 'replyCount', 'viewCount', 'hasNewPosts',
                      'isClosed', 'isSubscribed', 'canSubscribe', 'url', 'shortContent', 'participatedUserIds',
                      'isPinned', 'isAnnouncement', 'isStickySource', 'canRename', 'canDelete', 'canClose',
                      'canApprove', 'canStick', 'canMove', 'canMerge', 'canBan', 'canReply', 'canReport',
                      'canUpload', 'isBanned', 'isApproved', 'isDeleted', 'isMoved', 'isMerged', 'realTopicId',
                      'canLike', 'isLiked', 'likeCount', 'hasPoll', 'poll'];

        foreach ($topicProps as $prop) {
            if (property_exists($this, $prop)) {
                $data[$prop] = $this->$prop;
            }
        }
        $data['hasPoll'] = $this->hasPoll ?? false;
        $data['poll'] = $this->poll;

        return $data;
    }
}

/**
 * ForumCopilot Thread By Post Result
 * Maps from ThreadByPostData_Output
 */
class FCThreadByPostResult extends FCThreadResult
{
    public $canReply;
    public $canReport;
    public $canUpload;
    public $position;

    public function __construct($result = true, $resultText = 'success', $totalPostNum = 0, $canReply = false, $canReport = false, $canUpload = false, $posts = [], $position = 1, $topicData = [])
    {
        parent::__construct($result, $resultText, $totalPostNum, $posts, $topicData);
        $this->canReply = $canReply;
        $this->canReport = $canReport;
        $this->canUpload = $canUpload;
        $this->position = $position;
    }

    public function toArray()
    {
        $data = parent::toArray();
        $data['canReply'] = $this->canReply;
        $data['canReport'] = $this->canReport;
        $data['canUpload'] = $this->canUpload;
        $data['position'] = $this->position;
        return $data;
    }
}

File: packages/forumcopilot_sdk/lib/models/results/fc_post_result.dart
import 'package:dart_mappable/dart_mappable.dart';
import 'package:forumcopilot_sdk/models/entities/fc_attachment.dart';
import 'package:forumcopilot_sdk/models/entities/fc_attachment_data.dart';
import 'package:forumcopilot_sdk/models/entities/fc_post.dart';
import 'package:forumcopilot_sdk/models/entities/fc_poll.dart';
import 'package:forumcopilot_sdk/models/entities/fc_topic.dart';
import 'package:forumcopilot_sdk/models/mapping/hooks.dart';
import 'package:forumcopilot_sdk/models/results/fc_base_result.dart';
import 'package:forumcopilot_sdk/models/results/fc_topic_result.dart';

part 'fc_post_result.mapper.dart';

/// Forum Copilot Thread Result
/// Maps from ThreadData_Output
@MappableClass()
class FCThreadResult extends FCTopic with FCThreadResultMappable {
  /// Whether the operation was successful
  bool result;

  /// Result message or error text (only present when result = false)
  String? resultText;

  /// Total number of posts in this topic
  int totalPostNum;

  /// An array contains a list of posts
  List<FCPost> posts;

  FCThreadResult({
    required this.result,
    this.resultText,
    required this.totalPostNum,
    this.posts = const [],
    // FCTopic required fields
    required String id,
    required String title,
    required String forumId,
    required String forumName,
    required String authorId,
    required String authorName,
    required String authorUserType,
    required DateTime timestamp,
    // FCTopic optional fields
    String? prefix,
    String? authorIconUrl,
    int replyCount = 0,
    int viewCount = 0,
    bool hasNewPosts = false,
    bool isClosed = false,
    bool isSubscribed = false,
    bool canSubscribe = true,
    String? url,
    String? shortContent,
    List<String> participatedUserIds = const [],
    bool isPinned = false,
    bool isAnnouncement = false,
    bool isStickySource = false,
    // Moderation capabilities
    bool canRename = false,
    bool canDelete = false,
    bool canClose = false,
    bool canApprove = false,
    bool canStick = false,
    bool canMove = false,
    bool canMerge = false,
    bool canBan = false,
    // User interaction capabilities
    bool canReply = false,
    bool canReport = false,
    bool canUpload = false,
    // Moderation statuses
    bool isBanned = false,
    bool isApproved = true,
    bool isDeleted = false,
    bool isMoved = false,
    bool isMerged = false,
    String? realTopicId,
    // Like/thank
    bool canLike = false,
    bool isLiked = false,
    int likeCount = 0,
    bool canThank = false,
    bool hasPoll = false,
    FCPoll? poll,
  }) : super(
          id: id,
          title: title,
          forumId: forumId,
          forumName: forumName,
          authorId: authorId,
          authorName: authorName,
          authorUserType: authorUserType,
          timestamp: timestamp,
          prefix: prefix,
          authorIconUrl: authorIconUrl,
          replyCount: replyCount,
          viewCount: viewCount,
          hasNewPosts: hasNewPosts,
          isClosed: isClosed,
          isSubscribed: isSubscribed,
          canSubscribe: canSubscribe,
          url: url,
          shortContent: shortContent,
          participatedUserIds: participatedUserIds,
          isPinned: isPinned,
          isAnnouncement: isAnnouncement,
          isStickySource: isStickySource,
          canRename: canRename,
          canDelete: canDelete,
          canClose: canClose,
          canApprove: canApprove,
          canStick: canStick,
          canMove: canMove,
          canMerge: canMerge,
          canBan: canBan,
          isBanned: isBanned,
          isApproved: isApproved,
          isDeleted: isDeleted,
          isMoved: isMoved,
          isMerged: isMerged,
          realTopicId: realTopicId,
          canLike: canLike,
          isLiked: isLiked,
          likeCount: likeCount,
          canThank: canThank,
          hasPoll: hasPoll,
          poll: poll,
          canReply: canReply,
          canReport: canReport,
          canUpload: canUpload,
        );
}

/// Forum Copilot Thread By Unread Result
/// Maps from ThreadByUnreadData_Output
@MappableClass()
class FCThreadByUnreadResult extends FCTopic with FCThreadByUnreadResultMappable {
  /// Whether the operation was successful
  bool result;

  /// Result message or error text (only present when result = false)
  String? resultText;

  /// Total number of posts in this topic
  int totalPostNum;

  /// Return false if user cannot reply to this thread
  bool canReply;

  /// Return true if if user can report post in this thread
  bool canReport;

  /// Return false if user cannot attach files to this thread
  bool canUpload;

  /// An array contains a list of posts
  List<FCPost> posts;

  /// This value is to indicate the current first unread post position relative to the entire thread,
  /// starting from "1" to indicate first post, thereafter.
  int position;

  FCThreadByUnreadResult({
    required this.result,
    this.resultText,
    required this.totalPostNum,
    this.canReply = false,
    this.canReport = false,
    this.canUpload = false,
    this.posts = const [],
    required this.position,
    // FCTopic required fields
    required String id,
    required String title,
    required String forumId,
    required String forumName,
    required String authorId,
    required String authorName,
    required String authorUserType,
    required DateTime timestamp,
    // FCTopic optional fields
    String? prefix,
    String? authorIconUrl,
    int replyCount = 0,
    int viewCount = 0,
    bool hasNewPosts = false,
    bool isClosed = false,
    bool isSubscribed = false,
    bool canSubscribe = true,
    String? url,
    String? shortContent,
    List<String> participatedUserIds = const [],
    bool isPinned = false,
    bool isAnnouncement = false,
    bool isStickySource = false,
    // Moderation capabilities
    bool canRename = false,
    bool canDelete = false,
    bool canClose = false,
    bool canApprove = false,
    bool canStick = false,
    bool canMove = false,
    bool canMerge = false,
    bool canBan = false,
    // Moderation statuses
    bool isBanned = false,
    bool isApproved = true,
    bool isDeleted = false,
    bool isMoved = false,
    bool isMerged = false,
    String? realTopicId,
    // Like/thank
    bool canLike = false,
    bool isLiked = false,
    int likeCount = 0,
    bool canThank = false,
    bool hasPoll = false,
    FCPoll? poll,
  }) : super(
          id: id,
          title: title,
          forumId: forumId,
          forumName: forumName,
          authorId: authorId,
          authorName: authorName,
          authorUserType: authorUserType,
          timestamp: timestamp,
          prefix: prefix,
          authorIconUrl: authorIconUrl,
          replyCount: replyCount,
          viewCount: viewCount,
          hasNewPosts: hasNewPosts,
          isClosed: isClosed,
          isSubscribed: isSubscribed,
          canSubscribe: canSubscribe,
          url: url,
          shortContent: shortContent,
          participatedUserIds: participatedUserIds,
          isPinned: isPinned,
          isAnnouncement: isAnnouncement,
          isStickySource: isStickySource,
          canRename: canRename,
          canDelete: canDelete,
          canClose: canClose,
          canApprove: canApprove,
          canStick: canStick,
          canMove: canMove,
          canMerge: canMerge,
          canBan: canBan,
          isBanned: isBanned,
          isApproved: isApproved,
          isDeleted: isDeleted,
          isMoved: isMoved,
          isMerged: isMerged,
          realTopicId: realTopicId,
          canLike: canLike,
          isLiked: isLiked,
          likeCount: likeCount,
          canThank: canThank,
          hasPoll: hasPoll,
          poll: poll,
          canReply: canReply,
          canReport: canReport,
          canUpload: canUpload,
        );
}

/// Forum Copilot Thread By Post Result
/// Maps from ThreadByPostData_Output
@MappableClass()
class FCThreadByPostResult extends FCTopic with FCThreadByPostResultMappable {
  /// Whether the operation was successful
  bool result;

  /// Result message or error text (only present when result = false)
  String? resultText;

  /// Total number of posts in this topic
  int totalPostNum;

  /// Return false if user cannot reply to this thread
  bool canReply;

  /// Return true if if user can report post in this thread
  bool canReport;

  /// Return false if user cannot attach files to this thread
  bool canUpload;

  /// An array contains a list of posts
  List<FCPost> posts;

  /// This value is to indicate the current anchor post position relative to the entire thread,
  /// starting from "1" to indicate first post, thereafter.
  int position;

  FCThreadByPostResult({
    required this.result,
    this.resultText,
    required this.totalPostNum,
    this.canReply = false,
    this.canReport = false,
    this.canUpload = false,
    this.posts = const [],
    required this.position,
    // FCTopic required fields
    required String id,
    required String title,
    required String forumId,
    required String forumName,
    required String authorId,
    required String authorName,
    required String authorUserType,
    required DateTime timestamp,
    // FCTopic optional fields
    String? prefix,
    String? authorIconUrl,
    int replyCount = 0,
    int viewCount = 0,
    bool hasNewPosts = false,
    bool isClosed = false,
    bool isSubscribed = false,
    bool canSubscribe = true,
    String? url,
    String? shortContent,
    List<String> participatedUserIds = const [],
    bool isPinned = false,
    bool isAnnouncement = false,
    bool isStickySource = false,
    // Moderation capabilities
    bool canRename = false,
    bool canDelete = false,
    bool canClose = false,
    bool canApprove = false,
    bool canStick = false,
    bool canMove = false,
    bool canMerge = false,
    bool canBan = false,
    // Moderation statuses
    bool isBanned = false,
    bool isApproved = true,
    bool isDeleted = false,
    bool isMoved = false,
    bool isMerged = false,
    String? realTopicId,
    // Like/thank
    bool canLike = false,
    bool isLiked = false,
    int likeCount = 0,
    bool canThank = false,
    bool hasPoll = false,
    FCPoll? poll,
  }) : super(
          id: id,
          title: title,
          forumId: forumId,
          forumName: forumName,
          authorId: authorId,
          authorName: authorName,
          authorUserType: authorUserType,
          timestamp: timestamp,
          prefix: prefix,
          authorIconUrl: authorIconUrl,
          replyCount: replyCount,
          viewCount: viewCount,
          hasNewPosts: hasNewPosts,
          isClosed: isClosed,
          isSubscribed: isSubscribed,
          canSubscribe: canSubscribe,
          url: url,
          shortContent: shortContent,
          participatedUserIds: participatedUserIds,
          isPinned: isPinned,
          isAnnouncement: isAnnouncement,
          isStickySource: isStickySource,
          canRename: canRename,
          canDelete: canDelete,
          canClose: canClose,
          canApprove: canApprove,
          canStick: canStick,
          canMove: canMove,
          canMerge: canMerge,
          canBan: canBan,
          isBanned: isBanned,
          isApproved: isApproved,
          isDeleted: isDeleted,
          isMoved: isMoved,
          isMerged: isMerged,
          realTopicId: realTopicId,
          canLike: canLike,
          isLiked: isLiked,
          likeCount: likeCount,
          canThank: canThank,
          hasPoll: hasPoll,
          poll: poll,
          canReply: canReply,
          canReport: canReport,
          canUpload: canUpload,
        );
}

/// Forum Copilot Reply Post Result
/// Maps from ReplyPostData_Output
@MappableClass()
class FCReplyPostResult extends FCBaseResult with FCReplyPostResultMappable {
  /// The newly generated post ID for this new post
  String? postId;

  /// 1 = post is success but need moderation. Otherwise no need to return this key
  int state;

  /// Return the newly replied post content
  String? postContent;

  /// Return true if user can edit this post
  bool canEdit;

  /// Return true if user can delete this post
  bool canDelete;

  /// Return true if user can report this post
  bool canReport;

  FCReplyPostResult({
    required bool result,
    String? resultText,
    this.postId,
    this.state = 0,
    this.postContent,
    this.canEdit = false,
    this.canDelete = false,
    this.canReport = false,
  }) : super(result: result, resultText: resultText);
}

/// Forum Copilot Report Post Result
/// Maps from ReportPostData_Output
@MappableClass()
class FCReportPostResult extends FCBaseResult with FCReportPostResultMappable {
  FCReportPostResult({
    required bool result,
    String? resultText,
  }) : super(result: result, resultText: resultText);
}

/// Forum Copilot Quote Post Result
/// Maps from QuotePostData_Output
@MappableClass()
class FCQuotePostResult extends FCBaseResult with FCQuotePostResultMappable {
  /// Quoted post content
  String? quoteContent;

  FCQuotePostResult({
    required bool result,
    String? resultText,
    this.quoteContent,
  }) : super(result: result, resultText: resultText);
}

/// Forum Copilot Raw Post Result
/// Maps from RawPostData_Output
@MappableClass()
class FCRawPostResult extends FCBaseResult with FCRawPostResultMappable {
  /// Raw post content
  String? postContent;

  /// Post title (only for first post)
  String? postTitle;

  /// Whether the user can edit the title
  @MappableField(key: 'canEditTitle')
  bool? canEditTitle;

  /// Forum ID where the post belongs (always present for uploads)
  @MappableField(key: 'forumId')
  String? forumId;

  /// List of existing attachments
  List<FCAttachment>? attachments;

  /// Attachment editor metadata for managing attachments during edit
  @MappableField(key: 'attachmentData')
  FCAttachmentData? attachmentData;

  /// Current prefix ID of the thread (null if no prefix is set)
  /// Only present when editing the first post
  @MappableField(key: 'prefixId')
  String? prefixId;

  /// Whether a prefix is required for this thread
  /// Only true when editing the first post
  @MappableField(key: 'requirePrefix')
  bool requirePrefix;

  /// List of available prefixes for this forum
  /// Empty array for non-first posts or when no prefixes are available
  List<FCPrefix> prefixes;

  FCRawPostResult({
    required bool result,
    String? resultText,
    this.postContent,
    this.postTitle,
    this.canEditTitle,
    this.forumId,
    this.attachments,
    this.attachmentData,
    this.prefixId,
    this.requirePrefix = false,
    this.prefixes = const [],
  }) : super(result: result, resultText: resultText);
}

/// Forum Copilot Save Raw Post Result
/// Maps from SaveRawPostData_Output
@MappableClass()
class FCSaveRawPostResult extends FCBaseResult with FCSaveRawPostResultMappable {
  /// Updated post content
  String? postContent;

  FCSaveRawPostResult({
    required bool result,
    String? resultText,
    this.postContent,
  }) : super(result: result, resultText: resultText);
}

/// Forum Copilot Announcement Result
/// Maps from AnnouncementData_Output
@MappableClass()
class FCAnnouncementResult extends FCTopic with FCAnnouncementResultMappable {
  /// Whether the operation was successful
  bool result;

  /// Result message or error text (only present when result = false)
  String? resultText;

  /// Total number of posts in this topic
  int totalPostNum;

  /// Announcement content
  String? announcementContent;

  /// Announcement title
  String? announcementTitle;

  /// List of posts in this thread
  List<FCPost> posts;

  FCAnnouncementResult({
    required this.result,
    this.resultText,
    required this.totalPostNum,
    this.announcementContent,
    this.announcementTitle,
    this.posts = const [],
    // FCTopic required fields
    required String id,
    required String title,
    required String forumId,
    required String forumName,
    required String authorId,
    required String authorName,
    required String authorUserType,
    required DateTime timestamp,
    // FCTopic optional fields
    String? prefix,
    String? authorIconUrl,
    int replyCount = 0,
    int viewCount = 0,
    bool hasNewPosts = false,
    bool isClosed = false,
    bool isSubscribed = false,
    bool canSubscribe = true,
    String? url,
    String? shortContent,
    List<String> participatedUserIds = const [],
    bool isPinned = false,
    bool isAnnouncement = false,
    bool isStickySource = false,
    // Moderation capabilities
    bool canRename = false,
    bool canDelete = false,
    bool canClose = false,
    bool canApprove = false,
    bool canStick = false,
    bool canMove = false,
    bool canMerge = false,
    bool canBan = false,
    // User interaction capabilities
    bool canReply = false,
    bool canReport = false,
    bool canUpload = false,
    // Moderation statuses
    bool isBanned = false,
    bool isApproved = true,
    bool isDeleted = false,
    bool isMoved = false,
    bool isMerged = false,
    String? realTopicId,
    // Like/thank
    bool canLike = false,
    bool isLiked = false,
    int likeCount = 0,
    bool canThank = false,
    bool hasPoll = false,
    FCPoll? poll,
  }) : super(
          id: id,
          title: title,
          forumId: forumId,
          forumName: forumName,
          authorId: authorId,
          authorName: authorName,
          authorUserType: authorUserType,
          timestamp: timestamp,
          prefix: prefix,
          authorIconUrl: authorIconUrl,
          replyCount: replyCount,
          viewCount: viewCount,
          hasNewPosts: hasNewPosts,
          isClosed: isClosed,
          isSubscribed: isSubscribed,
          canSubscribe: canSubscribe,
          url: url,
          shortContent: shortContent,
          participatedUserIds: participatedUserIds,
          isPinned: isPinned,
          isAnnouncement: isAnnouncement,
          isStickySource: isStickySource,
          canRename: canRename,
          canDelete: canDelete,
          canClose: canClose,
          canApprove: canApprove,
          canStick: canStick,
          canMove: canMove,
          canMerge: canMerge,
          canBan: canBan,
          isBanned: isBanned,
          isApproved: isApproved,
          isDeleted: isDeleted,
          isMoved: isMoved,
          isMerged: isMerged,
          realTopicId: realTopicId,
          canLike: canLike,
          isLiked: isLiked,
          likeCount: likeCount,
          canThank: canThank,
          hasPoll: hasPoll,
          poll: poll,
          canReply: canReply,
          canReport: canReport,
          canUpload: canUpload,
        );
}
